pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/MultiConsumerPulsarSource.java

```diff
@@ -69,7 +72,20 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
             cb.messageListener(this);
 
             Consumer<T> consumer = cb.subscribeAsync().join();
-            inputConsumers.add(consumer);
+            inputConsumers.put(TopicName.get(topic), consumer);
+        }
+        if (sourceContext instanceof ExtendedSourceContext) {
```

Comment:
I think this is a little more complicated that it needs to be.  After the a PulsarSource is done opening, we can just get a list of consumers from the PulsarSource and pass it into the ContextImpl.  That approach will be much simpler and we don't need to the additional interfaces.

Reply:
One way or another the PulsarSource will need to set data to the context. 
In this case we either make pulsar source depend on concrete ContextImpl, or extend SourceContext with a setter that is not needed in actual Source, or add the interface that extends SourceContext to not make that method a part of the public API.
Whether it sets map of consumers or a func that resolves it is a matter of choice. I preferred the func to keep topic comparison logic in the PulsarSource plus case PulsarSource ever decides to create consumers dynamically/after open()

Reply:
@dlg99 In `JavaInstanceRunnable`, once you instantiate and call `open` on the PulsarSource, just also call another method called `getConsumers()` to get the list of consumers and then pass it into ContextImpl when it gets instantiated.  It is much simpler that way.